{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c9a615a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import requests\n",
    "from pprint import pprint\n",
    "import pandas as pd\n",
    "import bioregistry as br\n",
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6a8a3068",
   "metadata": {},
   "outputs": [],
   "source": [
    "url= 'https://smart-api.info/api/query?q=tags.name:translator&size=150&fields=info,paths,servers,tags,components.x-bte*,info,_meta'\n",
    "response = requests.get(url)\n",
    "data = []\n",
    "\n",
    "for api in response.json().get('hits'):\n",
    "    info = api.get('info')\n",
    "    \n",
    "    xtranslator = info.get('x-translator')\n",
    "    version = info.get('version')\n",
    "    if xtranslator is not None:\n",
    "        component = xtranslator.get('component')\n",
    "        team = xtranslator.get('team')\n",
    "        biolink_version = xtranslator.get('biolink-version')\n",
    "        team = xtranslator.get('team')\n",
    "    else:\n",
    "        component = \"\"\n",
    "        team = \"\"\n",
    "        biolink_version = \"\"\n",
    "        team = \"\"\n",
    "    \n",
    "    xtrapi = info.get('x-trapi')\n",
    "    if xtrapi is not None:\n",
    "        trapi_version = xtrapi.get('version')\n",
    "    else:\n",
    "        trapi_version = 'unknown'\n",
    "    title = info.get('title')\n",
    "    \n",
    "    description = info.get('description')\n",
    "    contact = info.get('contact')\n",
    "    contact_name = contact.get('name')\n",
    "    contact_email = contact.get('email')\n",
    "    servers = api.get('servers')\n",
    "    api_id = api.get('_id')\n",
    "    #servers = api.get('servers')\n",
    "    #if servers is not None:\n",
    "    #    for server in servers:\n",
    "    #        server_url = server.get('url')\n",
    "    #        server_description = server.get('description')\n",
    "    #        row = {\n",
    "    #            \"title\": title,\n",
    "    #            \"component\": component,\n",
    "    #            \"description\": description,\n",
    "    #            \"team\":team,\n",
    "    #            \"trapi_version\":trapi_version,\n",
    "    #            \"biolink_version\":biolink_version,\n",
    "    #            \"contact_name\":contact_name,\n",
    "    #            \"contact_email\":contact_email,\n",
    "    #            \"server_url\": server_url,\n",
    "    #            \"server_description\": server_description\n",
    "    #        }\n",
    "    #        data.append(row)\n",
    "    #else:\n",
    "    #    server_url = \"unknonwn\"\n",
    "    #    server_description = \"unknown\"\n",
    "\n",
    "    row = {\n",
    "                \"smart_api_url\": \"http://smart-api.info/registry?q=\"+api_id,\n",
    "                \"title\": title,\n",
    "                \"component\": component,\n",
    "                \"description\": description,\n",
    "                \"team\":team,\n",
    "                \"version\":version,\n",
    "                \"trapi_version\":trapi_version,\n",
    "                \"biolink_version\":biolink_version,\n",
    "                \"contact_name\":contact_name,\n",
    "                \"contact_email\":contact_email,\n",
    "                \"servers\":servers\n",
    "    }\n",
    "    data.append(row)\n",
    "    \n",
    "#df = pd.DataFrame (data, columns = ['title','component','team','trapi_version','biolink_version','contact_name', 'contact_email', 'server_url', 'server_description'])\n",
    "df_description = pd.DataFrame (data, columns = ['smart_api_url','title','component','description',\n",
    "                                                'team','version','trapi_version','biolink_version',\n",
    "                                                'contact_name', 'contact_email', 'servers'])\n",
    "#df.to_csv(r'apis.csv', index = False)\n",
    "df_description.to_csv(r'apis_with_description.csv', index = False)\n",
    "# print(df)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e3de89a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trusted resource not found for: ac\n",
      "trusted resource not found for: afo\n",
      "trusted resource not found for: alzforum.mutation\n",
      "trusted resource not found for: araport\n",
      "trusted resource not found for: asrp\n",
      "trusted resource not found for: atol\n",
      "trusted resource not found for: bactibase\n",
      "trusted resource not found for: bams\n",
      "trusted resource not found for: bcio\n",
      "trusted resource not found for: bel\n",
      "trusted resource not found for: bioregistry\n",
      "trusted resource not found for: bioregistry.collection\n",
      "trusted resource not found for: biorxiv\n",
      "trusted resource not found for: birnlex\n",
      "trusted resource not found for: bmrb.restraint\n",
      "trusted resource not found for: bpdb\n",
      "trusted resource not found for: bs\n",
      "trusted resource not found for: caloha\n",
      "trusted resource not found for: cba\n",
      "trusted resource not found for: cgnc\n",
      "trusted resource not found for: citexplore\n",
      "trusted resource not found for: clingene\n",
      "trusted resource not found for: cmpo\n",
      "trusted resource not found for: co_320\n",
      "trusted resource not found for: co_321\n",
      "trusted resource not found for: co_321:root\n",
      "trusted resource not found for: co_322\n",
      "trusted resource not found for: co_323\n",
      "trusted resource not found for: co_324\n",
      "trusted resource not found for: co_325\n",
      "trusted resource not found for: co_327\n",
      "trusted resource not found for: co_330\n",
      "trusted resource not found for: co_331\n",
      "trusted resource not found for: co_333\n",
      "trusted resource not found for: co_334\n",
      "trusted resource not found for: co_335\n",
      "trusted resource not found for: co_336\n",
      "trusted resource not found for: co_337\n",
      "trusted resource not found for: co_338\n",
      "trusted resource not found for: co_339\n",
      "trusted resource not found for: co_340\n",
      "trusted resource not found for: co_341\n",
      "trusted resource not found for: co_343\n",
      "trusted resource not found for: co_345\n",
      "trusted resource not found for: co_346\n",
      "trusted resource not found for: co_347\n",
      "trusted resource not found for: co_348\n",
      "trusted resource not found for: co_350\n",
      "trusted resource not found for: co_356\n",
      "trusted resource not found for: co_357\n",
      "trusted resource not found for: co_358\n",
      "trusted resource not found for: co_359\n",
      "trusted resource not found for: co_360\n",
      "trusted resource not found for: co_365\n",
      "trusted resource not found for: co_366\n",
      "trusted resource not found for: cohd\n",
      "trusted resource not found for: come\n",
      "trusted resource not found for: conso\n",
      "trusted resource not found for: covoc\n",
      "trusted resource not found for: cp\n",
      "trusted resource not found for: cryoem\n",
      "trusted resource not found for: csp\n",
      "trusted resource not found for: dc\n",
      "trusted resource not found for: dcat\n",
      "trusted resource not found for: dcterms\n",
      "trusted resource not found for: dctypes\n",
      "trusted resource not found for: dermo\n",
      "trusted resource not found for: dicom\n",
      "trusted resource not found for: dictybase\n",
      "trusted resource not found for: diseasesdb\n",
      "trusted resource not found for: drugbank.salt\n",
      "trusted resource not found for: drugcentral\n",
      "trusted resource not found for: dto\n",
      "trusted resource not found for: ecacc\n",
      "trusted resource not found for: ecg\n",
      "trusted resource not found for: ecmdb\n",
      "trusted resource not found for: ecocyc\n",
      "trusted resource not found for: empiar\n",
      "trusted resource not found for: enm\n",
      "trusted resource not found for: ensemblglossary\n",
      "trusted resource not found for: eol\n",
      "trusted resource not found for: epcc\n",
      "trusted resource not found for: eurofir\n",
      "trusted resource not found for: evm\n",
      "trusted resource not found for: faldo\n",
      "trusted resource not found for: fbql\n",
      "trusted resource not found for: fbtc\n",
      "trusted resource not found for: fishbase.species\n",
      "trusted resource not found for: foaf\n",
      "trusted resource not found for: fyler\n",
      "trusted resource not found for: galen\n",
      "trusted resource not found for: gard\n",
      "trusted resource not found for: gbif\n",
      "trusted resource not found for: gc\n",
      "trusted resource not found for: genbank\n",
      "trusted resource not found for: gexo\n",
      "trusted resource not found for: ghr\n",
      "trusted resource not found for: glygen\n",
      "trusted resource not found for: gmelin\n",
      "trusted resource not found for: go.model\n",
      "trusted resource not found for: goeco\n",
      "trusted resource not found for: google.scholar\n",
      "trusted resource not found for: gorel\n",
      "trusted resource not found for: gsfa\n",
      "trusted resource not found for: gtr\n",
      "trusted resource not found for: hcao\n",
      "trusted resource not found for: hcpcs\n",
      "trusted resource not found for: hms.lincs.compound\n",
      "trusted resource not found for: hog\n",
      "trusted resource not found for: hpath\n",
      "trusted resource not found for: hsdb\n",
      "trusted resource not found for: icd11\n",
      "trusted resource not found for: icd9\n",
      "trusted resource not found for: icdo\n",
      "trusted resource not found for: idocovid19\n",
      "trusted resource not found for: idoden\n",
      "trusted resource not found for: imdrf\n",
      "trusted resource not found for: inn\n",
      "trusted resource not found for: iro\n",
      "trusted resource not found for: itis\n",
      "trusted resource not found for: jax\n",
      "trusted resource not found for: kegg.enzyme\n",
      "trusted resource not found for: kegg.ligand\n",
      "trusted resource not found for: kupo\n",
      "trusted resource not found for: lbo\n",
      "trusted resource not found for: lncipedia\n",
      "trusted resource not found for: loinc\n",
      "trusted resource not found for: lspci\n",
      "trusted resource not found for: matrixdb\n",
      "trusted resource not found for: mcc\n",
      "trusted resource not found for: miaa\n",
      "trusted resource not found for: mirbase.family\n",
      "trusted resource not found for: mirte\n",
      "trusted resource not found for: mmsl\n",
      "trusted resource not found for: molbase.sheffield\n",
      "trusted resource not found for: msigdb\n",
      "trusted resource not found for: msio\n",
      "trusted resource not found for: namerxn\n",
      "trusted resource not found for: ncats.drug\n",
      "trusted resource not found for: nddf\n",
      "trusted resource not found for: ndex\n",
      "trusted resource not found for: ndfrt\n",
      "trusted resource not found for: neuronames\n",
      "trusted resource not found for: nextprot.family\n",
      "trusted resource not found for: nist\n",
      "trusted resource not found for: nlm\n",
      "trusted resource not found for: nlxanat\n",
      "trusted resource not found for: nlxdys\n",
      "trusted resource not found for: npass\n",
      "trusted resource not found for: oa\n",
      "trusted resource not found for: obo\n",
      "trusted resource not found for: oboinowl\n",
      "trusted resource not found for: om\n",
      "trusted resource not found for: omop\n",
      "trusted resource not found for: oncotree\n",
      "trusted resource not found for: orth\n",
      "trusted resource not found for: owl\n",
      "trusted resource not found for: panorama\n",
      "trusted resource not found for: pathbank\n",
      "trusted resource not found for: pdbe\n",
      "trusted resource not found for: peff\n",
      "trusted resource not found for: pesticides\n",
      "trusted resource not found for: pfam.clan\n",
      "trusted resource not found for: pfr\n",
      "trusted resource not found for: phi\n",
      "trusted resource not found for: pictar\n",
      "trusted resource not found for: ppdb\n",
      "trusted resource not found for: ppr\n",
      "trusted resource not found for: prov\n",
      "trusted resource not found for: pseudogene\n",
      "trusted resource not found for: pspub\n",
      "trusted resource not found for: publons.researcher\n",
      "trusted resource not found for: qudt\n",
      "trusted resource not found for: rdf\n",
      "trusted resource not found for: rdfs\n",
      "trusted resource not found for: rdo\n",
      "trusted resource not found for: reaxys\n",
      "trusted resource not found for: receptome.family\n",
      "trusted resource not found for: reproduceme\n",
      "trusted resource not found for: reto\n",
      "trusted resource not found for: rexo\n",
      "trusted resource not found for: rnamod\n",
      "trusted resource not found for: rxnorm\n",
      "trusted resource not found for: sael\n",
      "trusted resource not found for: salk\n",
      "trusted resource not found for: scopus\n",
      "trusted resource not found for: sdgio\n",
      "trusted resource not found for: sio\n",
      "trusted resource not found for: skos\n",
      "trusted resource not found for: smid\n",
      "trusted resource not found for: smiles\n",
      "trusted resource not found for: snap\n",
      "trusted resource not found for: snornabase\n",
      "trusted resource not found for: span\n",
      "trusted resource not found for: srao\n",
      "trusted resource not found for: ssbd.dataset\n",
      "trusted resource not found for: ssbd.project\n",
      "trusted resource not found for: sty\n",
      "trusted resource not found for: syoid\n",
      "trusted resource not found for: te\n",
      "trusted resource not found for: tfclass\n",
      "trusted resource not found for: th\n",
      "trusted resource not found for: ubprop\n",
      "trusted resource not found for: ucsc\n",
      "trusted resource not found for: uniprot.disease\n",
      "trusted resource not found for: uniprot.keyword\n",
      "trusted resource not found for: uniprot.location\n",
      "trusted resource not found for: uniprot.var\n",
      "trusted resource not found for: unirule\n",
      "trusted resource not found for: unpd\n",
      "trusted resource not found for: vandf\n",
      "trusted resource not found for: vega\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trusted resource not found for: vido\n",
      "trusted resource not found for: vmc\n",
      "trusted resource not found for: void\n",
      "trusted resource not found for: vsdb\n",
      "trusted resource not found for: vuid\n",
      "trusted resource not found for: vz\n",
      "trusted resource not found for: webelements\n",
      "trusted resource not found for: wgs84\n",
      "trusted resource not found for: wikidata.property\n",
      "trusted resource not found for: wormbase\n",
      "trusted resource not found for: xl\n",
      "trusted resource not found for: xsd\n",
      "trusted resource not found for: xuo\n",
      "trusted resource not found for: ymdb\n",
      "total number of resources % 1161\n"
     ]
    }
   ],
   "source": [
    "url= 'https://raw.githubusercontent.com/bioregistry/bioregistry/main/src/bioregistry/data/bioregistry.json'\n",
    "response = requests.get(url)\n",
    "data = []\n",
    "count = 0\n",
    "def make_row(api, data, prefix):\n",
    "    homepage = api.get('homepage')\n",
    "    name = api.get('name')\n",
    "    description = api.get('description')\n",
    "    row = {\n",
    "                \"short name\": prefix,\n",
    "                \"homepage\": homepage,\n",
    "                \"name\": name,\n",
    "                \"description\": description\n",
    "    }\n",
    "    data.append(row)\n",
    "    return data\n",
    "    \n",
    "for api in response.json():\n",
    "    count = count + 1\n",
    "    prefix_collection = response.json().get(api)\n",
    "    miriam = prefix_collection.get('miriam')\n",
    "    obofoundary = prefix_collection.get('obofoundry')\n",
    "    ols = prefix_collection.get('ols')\n",
    "    n2t = prefix_collection.get('n2t')\n",
    "    prefixcommons = prefix_collection.get('prefixcommons')\n",
    "    if miriam:\n",
    "        data = make_row(miriam, data, api)\n",
    "    elif obofoundary:\n",
    "        data = make_row(obofoundary, data, api)\n",
    "    elif n2t:\n",
    "        data = make_row(n2t, data, api)\n",
    "    else:\n",
    "        print(\"trusted resource not found for: \" + api)\n",
    "        \n",
    "print('total number of resources %', count)\n",
    "df_description = pd.DataFrame (data, columns = ['short name',\n",
    "                                     'homepage',\n",
    "                                     'name',\n",
    "                                     'description'\n",
    "                                    ])\n",
    "#df.to_csv(r'apis.csv', index = False)\n",
    "df_description.to_csv(r'miriam_registry_prefixes.csv', index = False)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b906028f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mismatch: bioregistry = 'clinicaltrials' infores = 'clinical-trials'\n",
      "mismatch: bioregistry = 'ncbigene' infores = 'entrez'\n",
      "mismatch: bioregistry = 'go' infores = 'go-bp'\n",
      "mismatch: bioregistry = 'go' infores = 'go-cc'\n",
      "mismatch: bioregistry = 'go' infores = 'go-mf'\n",
      "mismatch: bioregistry = 'lincs.smallmolecule' infores = 'lincs'\n",
      "mismatch: bioregistry = 'metacyc.compound' infores = 'metacyc'\n",
      "mismatch: bioregistry = 'ncbigene' infores = 'ncbi-gene'\n",
      "mismatch: bioregistry = 'mim' infores = 'omim'\n",
      "mismatch: bioregistry = 'orphanet.ordo' infores = 'ordo'\n",
      "mismatch: bioregistry = 'pathwaycommons' infores = 'pathway-commons'\n",
      "mismatch: bioregistry = 'pubchem.compound' infores = 'pubchem'\n",
      "mismatch: bioregistry = 'snomedct' infores = 'snomed'\n",
      "mismatch: bioregistry = 'so' infores = 'so '\n",
      "mismatch: bioregistry = 'wikidata' infores = 'wd'\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "import requests\n",
    "from pprint import pprint\n",
    "import pandas as pd\n",
    "import bioregistry\n",
    "import csv \n",
    "data = []\n",
    "registry = bioregistry.read_registry()\n",
    "\n",
    "\n",
    "with open('infores.tsv', newline = '') as infores:                                                                                          \n",
    "    infores_reader = csv.reader(infores, delimiter='\\t')\n",
    "    for infores in infores_reader:\n",
    "        br_name = bioregistry.normalize_prefix(infores[5])\n",
    "        if br_name is None:\n",
    "            continue\n",
    "        elif br_name != infores[5]:\n",
    "            print (\"mismatch: bioregistry = \" + \"'\" + br_name + \"'\" +\" infores = \" + \"'\" + infores[5] + \"'\"  )\n",
    "        else:\n",
    "            continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f283c21a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import requests\n",
    "from pprint import pprint\n",
    "import pandas as pd\n",
    "import bioregistry\n",
    "import csv \n",
    "data = []\n",
    "registry = bioregistry.read_registry()\n",
    "\n",
    "\n",
    "with open('infores.tsv', newline = '') as infores:                                                                                          \n",
    "    infores_reader = csv.reader(infores, delimiter='\\t')\n",
    "    for infores in infores_reader:\n",
    "        print(infores)\n",
    "        br_name = bioregistry.normalize_prefix(infores[5])\n",
    "        if br_name = '':\n",
    "            print \"no match for infores: \" + infores[5]\n",
    "        else:    \n",
    "            print (br_name + \" \" + infores[5])\n",
    "for registry_entry in registry:\n",
    "    homepage = bioregistry.get_homepage(registry_entry)\n",
    "    name = bioregistry.get_name(registry_entry)\n",
    "    description = bioregistry.get_description(registry_entry)\n",
    "    version = bioregistry.get_version(registry_entry)\n",
    "    if version:\n",
    "        infores = \"infores:\"+registry_entry+\".\"+version\n",
    "    else:\n",
    "        infores = \"infores:\"+registry_entry\n",
    "    row = {\n",
    "        \"short name\": registry_entry,\n",
    "        \"homepage\": homepage,\n",
    "        \"name\": name,\n",
    "        \"description\": description,\n",
    "        \"version\": version,\n",
    "        \"infores\": infores\n",
    "    }\n",
    "    data.append(row)\n",
    "    \n",
    "df_description = pd.DataFrame (data, columns = ['short name',\n",
    "                                     'homepage',\n",
    "                                     'name',\n",
    "                                     'description',\n",
    "                                     'version',\n",
    "                                     'infores'\n",
    "                                    ])\n",
    "#df.to_csv(r'apis.csv', index = False)\n",
    "csvdf_description.to_csv(r'miriam_registry_prefixes.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "2dabc65c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:ToolkitGenerator:Prefix case mismatch - supplied: Xenbase expected: XENBASE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44\n",
      "https://icees.renci.org:16339/meta_knowledge_graph\n",
      "https://unsecret.ncats.io/v2/meta_knowledge_graph\n",
      "https://icees.renci.org:16341/meta_knowledge_graph\n",
      "https://cam-kp-api-dev.renci.org/meta_knowledge_graph\n",
      "https://stars-app.renci.org/sparql-kp/meta_knowledge_graph\n",
      "https://automat.renci.org/hetio/1.1/meta_knowledge_graph\n",
      "https://automat.renci.org/human-goa/1.1/meta_knowledge_graph\n",
      "https://automat.renci.org/covidkopkg/1.1/meta_knowledge_graph\n",
      "https://automat.renci.org/mychem/1.1/meta_knowledge_graph\n",
      "https://automat.renci.org/cord19-scibite/1.1/meta_knowledge_graph\n",
      "https://automat.renci.org/gtex/1.1/meta_knowledge_graph\n",
      "https://automat.renci.org/viral-proteome/1.1/meta_knowledge_graph\n",
      "https://automat.renci.org/uberongraph/1.1/meta_knowledge_graph\n",
      "https://automat.renci.org/hgnc/1.1/meta_knowledge_graph\n",
      "https://automat.renci.org/ontological-hierarchy/1.1/meta_knowledge_graph\n",
      "https://automat.renci.org/ctd/1.1/meta_knowledge_graph\n",
      "https://automat.renci.org/biolink/1.1/meta_knowledge_graph\n",
      "https://automat.renci.org/foodb/1.1/meta_knowledge_graph\n",
      "https://automat.renci.org/hmdb/1.1/meta_knowledge_graph\n",
      "https://automat.renci.org/textminingkp/1.1/meta_knowledge_graph\n",
      "https://automat.renci.org/panther/1.1/meta_knowledge_graph\n",
      "https://automat.renci.org/chemical-normalization/1.1/meta_knowledge_graph\n",
      "https://automat.renci.org/kegg/1.1/meta_knowledge_graph\n",
      "https://automat.renci.org/cord19-scigraph/1.1/meta_knowledge_graph\n",
      "https://automat.renci.org/pharos/1.1/meta_knowledge_graph\n",
      "https://automat.renci.org/robokopkg/1.1/meta_knowledge_graph\n",
      "https://automat.renci.org/gtopdb/1.1/meta_knowledge_graph\n",
      "https://automat.renci.org/covid-phenotypes/1.1/meta_knowledge_graph\n",
      "https://automat.renci.org/chembio/1.1/meta_knowledge_graph\n",
      "https://automat.renci.org/intact/1.1/meta_knowledge_graph\n",
      "https://spokekp.healthdatascience.cloud/api/v1.1//meta_knowledge_graph\n",
      "https://ia.healthdatascience.cloud/api/v1.1//meta_knowledge_graph\n",
      "https://translator.broadinstitute.org/molepro/trapi/v1.1/meta_knowledge_graph\n",
      "https://explanatory-agent.azurewebsites.net/v1.1//meta_knowledge_graph\n",
      "https://openpredict.semanticscience.org/meta_knowledge_graph\n",
      "http://chp.thayer.dartmouth.edu//meta_knowledge_graph\n",
      "http://breast.chp.thayer.dartmouth.edu//meta_knowledge_graph\n",
      "http://brain.chp.thayer.dartmouth.edu//meta_knowledge_graph\n",
      "http://lung.chp.thayer.dartmouth.edu//meta_knowledge_graph\n",
      "https://arax.ncats.io/api/rtxkg2/v1.1/meta_knowledge_graph\n",
      "https://translator.broadinstitute.org/genetics_provider/trapi/v1.1/meta_knowledge_graph\n",
      "https://arax.ncats.io/api/arax/v1.1/meta_knowledge_graph\n",
      "https://cohd.io/api//meta_knowledge_graph\n",
      "https://covid.cohd.io/api//meta_knowledge_graph\n",
      "biolink:has_attribute\n",
      "biolink:negatively_regulates\n",
      "biolink:positively_regulates\n",
      "biolink:regulates\n",
      "biolink:type\n",
      "biolink:negatively_regulates_entity_to_entity\n",
      "biolink:positively_regulates_entity_to_entity\n",
      "biolink:in_clinical_trial_for\n",
      "biolink:negatively_regulates__entity_to_entity\n",
      "biolink:positively_regulates__entity_to_entity\n",
      "biolink:disruptive_inframe_deletion\n",
      "biolink:disruptive_inframe_insertion\n",
      "biolink:conservative_inframe_insertion\n",
      "biolink:conservative_inframe_deletion\n",
      "biolink:non_coding_transcript_variant\n",
      "biolink:regulates_process_to_process\n",
      "biolink:negatively_regulates_process_to_process\n",
      "biolink:associated_with_decreased_risk_for\n",
      "biolink:associated_with_increased_risk_for\n",
      "biolink:associated_with_risk_for\n",
      "biolink:has_evidence\n",
      "biolink:enriched_in_genes_correlated_with\n",
      "biolink:gene_has_variant_that_contributes_to_drug_response_association\n",
      "biolink:gene_has_variant_that_contributes_to_disease_association\n",
      "is_affected_by\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "import requests\n",
    "from pprint import pprint\n",
    "import pandas as pd\n",
    "import bioregistry as br\n",
    "import csv\n",
    "import bmt\n",
    "\n",
    "tk = bmt.Toolkit()\n",
    "url= 'https://smart-api.info/api/query?q=tags.name:translator&size=150&fields=info,servers,paths'\n",
    "response = requests.get(url)\n",
    "data = []\n",
    "count = 0\n",
    "for api in response.json().get('hits'):\n",
    "    info = api.get('info')\n",
    "    paths = api.get('paths') \n",
    "    title = info.get('title')\n",
    "    api_id = api.get('_id')\n",
    "    servers = api.get('servers')\n",
    "    if servers is not None:\n",
    "        for server in servers:\n",
    "            for path in paths:\n",
    "                if path == '/meta_knowledge_graph':\n",
    "                #if path == '/predicates':\n",
    "                    count = count + 1\n",
    "                    data.append(server.get('url')+path)\n",
    "print(count)\n",
    "predicates = []\n",
    "for meta_graph in data:\n",
    "    print(meta_graph)\n",
    "    response = requests.get(meta_graph)\n",
    "    if response.status_code != 404:\n",
    "        # pprint(response.json())\n",
    "        for edge in response.json().get('edges'):\n",
    "            predicate = edge.get('predicate')\n",
    "            if predicate not in predicates:\n",
    "                predicates.append(predicate)\n",
    "\n",
    "for predicate in predicates:\n",
    "    if tk.is_predicate(predicate):\n",
    "        continue\n",
    "        #print(predicate)\n",
    "    else:\n",
    "        print(predicate)\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "97e6e71a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:ToolkitGenerator:Prefix case mismatch - supplied: Xenbase expected: XENBASE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49\n",
      "biolink:association_slot\n",
      "biolink:associated_environmental_context|association slot\n",
      "biolink:has_confidence_level|association slot\n",
      "biolink:clinical_modifier_qualifier|association slot\n",
      "biolink:supporting_data_source|association slot\n",
      "biolink:logical_interpretation|association slot\n",
      "biolink:onset_qualifier|association slot\n",
      "biolink:stage_qualifier|association slot\n",
      "biolink:predicate|association slot\n",
      "biolink:knowledge_source|association slot\n",
      "biolink:primary_knowledge_source|knowledge source\n",
      "biolink:original_knowledge_source|primary knowledge source\n",
      "biolink:aggregator_knowledge_source|knowledge source\n",
      "biolink:reaction_direction|association slot\n",
      "biolink:qualifiers|association slot\n",
      "biolink:reaction_side|association slot\n",
      "biolink:negated|association slot\n",
      "biolink:subject|association slot\n",
      "biolink:chi_squared_statistic|association slot\n",
      "biolink:catalyst_qualifier|association slot\n",
      "biolink:has_evidence|association slot\n",
      "biolink:interacting_molecules_category|association slot\n",
      "biolink:provided_by|association slot\n",
      "biolink:reaction_balanced|association slot\n",
      "biolink:quantifier_qualifier|association slot\n",
      "biolink:p_value|association slot\n",
      "biolink:relation|association slot\n",
      "biolink:object|association slot\n",
      "biolink:phenotypic_state|association slot\n",
      "biolink:expression_site|association slot\n",
      "biolink:sequence_localization_attribute|association slot\n",
      "biolink:interbase_coordinate|sequence localization attribute\n",
      "biolink:start_interbase_coordinate|interbase coordinate\n",
      "biolink:end_interbase_coordinate|interbase coordinate\n",
      "biolink:phase|sequence localization attribute\n",
      "biolink:genome_build|sequence localization attribute\n",
      "biolink:base_coordinate|sequence localization attribute\n",
      "biolink:start_coordinate|base coordinate\n",
      "biolink:end_coordinate|base coordinate\n",
      "biolink:strand|sequence localization attribute\n",
      "biolink:sex_qualifier|association slot\n",
      "biolink:has_population_context|association slot\n",
      "biolink:stoichiometry|association slot\n",
      "biolink:severity_qualifier|association slot\n",
      "biolink:sequence_variant_qualifier|association slot\n",
      "biolink:has_temporal_context|association slot\n",
      "biolink:association_type|association slot\n",
      "biolink:publications|association slot\n",
      "biolink:frequency_qualifier|association slot\n",
      "biolink:has_attribute\n",
      "biolink:id\n",
      "biolink:iri\n",
      "biolink:type\n",
      "biolink:category|type\n",
      "biolink:name\n",
      "biolink:source\n",
      "biolink:description\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "import requests\n",
    "from pprint import pprint\n",
    "import pandas as pd\n",
    "import bioregistry as br\n",
    "import csv\n",
    "import bmt\n",
    "\n",
    "tk = bmt.Toolkit()\n",
    "edge_properties = []\n",
    "for edge_property in tk.get_descendants('association slot', True, True):\n",
    "    if edge_property not in edge_properties:\n",
    "        edge_properties.append(edge_property)\n",
    "print(len(edge_properties))\n",
    "for edge_property in tk.get_all_edge_properties(True):\n",
    "    if edge_property not in edge_properties:\n",
    "        edge_properties.append(edge_property)\n",
    "for prop in edge_properties:\n",
    "    #if (tk.get_element(prop).description is not None):\n",
    "    #    print = prop+'|'+(tk.get_element(prop).description)\n",
    "    if (tk.get_parent(prop) is not None):\n",
    "        print(prop+'|'+(tk.get_parent(prop)))\n",
    "    else:\n",
    "        print(prop)\n",
    "    \n",
    "# print(len(edge_properties))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "0529971a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "biolink:Association|id,iri,name,description,source,provided by,has attribute,subject,predicate,object,relation,negated,qualifiers,publications,association_type,association_category\n",
      "biolink:BehaviorToBehavioralFeatureAssociation|id,iri,name,description,source,provided by,has attribute,predicate,relation,negated,qualifiers,publications,association_type,association_category,behavior to behavioral feature association_subject,behavior to behavioral feature association_object,frequency qualifier,severity qualifier,onset qualifier,sex qualifier\n",
      "biolink:GenotypeToGenotypePartAssociation|id,iri,name,description,source,provided by,has attribute,relation,negated,qualifiers,publications,association_type,association_category,genotype to genotype part association_predicate,genotype to genotype part association_subject,genotype to genotype part association_object\n",
      "biolink:GeneRegulatoryRelationship|id,iri,name,description,source,provided by,has attribute,relation,negated,qualifiers,publications,association_type,association_category,gene regulatory relationship_predicate,gene regulatory relationship_subject,gene regulatory relationship_object\n",
      "biolink:NamedThingToInformationContentEntityAssociation|id,iri,name,description,source,provided by,has attribute,relation,negated,qualifiers,publications,association_type,association_category,named thing to information content entity association_subject,named thing to information content entity association_object,named thing to information content entity association_predicate\n",
      "biolink:DrugToGeneAssociation|id,iri,name,description,source,provided by,has attribute,subject,predicate,relation,negated,qualifiers,publications,association_type,association_category,drug to gene association_object\n",
      "biolink:VariantToPhenotypicFeatureAssociation|id,iri,name,description,source,provided by,has attribute,predicate,object,relation,negated,qualifiers,publications,association_type,association_category,variant to phenotypic feature association_subject,frequency qualifier,severity qualifier,onset qualifier,sex qualifier\n",
      "biolink:DiseaseToPhenotypicFeatureAssociation|id,iri,name,description,source,provided by,has attribute,subject,predicate,object,relation,negated,qualifiers,publications,association_type,association_category,frequency qualifier,severity qualifier,onset qualifier,sex qualifier\n",
      "biolink:OrganismToOrganismAssociation|id,iri,name,description,source,provided by,has attribute,predicate,negated,qualifiers,publications,association_type,association_category,organism to organism association_subject,organism to organism association_relation,organism to organism association_object\n",
      "biolink:OrganismTaxonToOrganismTaxonAssociation|id,iri,name,description,source,provided by,has attribute,predicate,relation,negated,qualifiers,publications,association_type,association_category,organism taxon to organism taxon association_subject,organism taxon to organism taxon association_object\n",
      "biolink:OrganismTaxonToOrganismTaxonSpecialization|id,iri,name,description,source,provided by,has attribute,relation,negated,qualifiers,publications,association_type,association_category,organism taxon to organism taxon specialization_subject,organism taxon to organism taxon specialization_object,organism taxon to organism taxon specialization_predicate\n",
      "biolink:OrganismTaxonToOrganismTaxonInteraction|id,iri,name,description,source,provided by,has attribute,relation,negated,qualifiers,publications,association_type,association_category,organism taxon to organism taxon interaction_associated environmental context,organism taxon to organism taxon interaction_subject,organism taxon to organism taxon interaction_object,organism taxon to organism taxon interaction_predicate\n",
      "biolink:CaseToPhenotypicFeatureAssociation|id,iri,name,description,source,provided by,has attribute,subject,predicate,object,relation,negated,qualifiers,publications,association_type,association_category,frequency qualifier,severity qualifier,onset qualifier,sex qualifier\n",
      "biolink:GenotypeToGeneAssociation|id,iri,name,description,source,provided by,has attribute,relation,negated,qualifiers,publications,association_type,association_category,genotype to gene association_predicate,genotype to gene association_subject,genotype to gene association_object\n",
      "biolink:SequenceVariantModulatesTreatmentAssociation|id,iri,name,description,source,provided by,has attribute,predicate,relation,negated,qualifiers,publications,association_type,association_category,sequence variant modulates treatment association_subject,sequence variant modulates treatment association_object\n",
      "biolink:SequenceAssociation|id,iri,name,description,source,provided by,has attribute,subject,predicate,object,relation,negated,qualifiers,publications,association_type,association_category\n",
      "biolink:GenomicSequenceLocalization|id,iri,name,description,source,provided by,has attribute,relation,negated,qualifiers,publications,association_type,association_category,start interbase coordinate,end interbase coordinate,genome build,strand,phase,genomic sequence localization_subject,genomic sequence localization_object,genomic sequence localization_predicate\n",
      "biolink:ChemicalToDiseaseOrPhenotypicFeatureAssociation|id,iri,name,description,source,provided by,has attribute,subject,predicate,relation,negated,qualifiers,publications,association_type,association_category,chemical to disease or phenotypic feature association_object\n",
      "biolink:GenotypeToVariantAssociation|id,iri,name,description,source,provided by,has attribute,relation,negated,qualifiers,publications,association_type,association_category,genotype to variant association_predicate,genotype to variant association_subject,genotype to variant association_object\n",
      "biolink:TaxonToTaxonAssociation|id,iri,name,description,source,provided by,has attribute,predicate,negated,qualifiers,publications,association_type,association_category,taxon to taxon association_subject,taxon to taxon association_relation,taxon to taxon association_object\n",
      "biolink:GeneToExpressionSiteAssociation|id,iri,name,description,source,provided by,has attribute,relation,negated,qualifiers,publications,association_type,association_category,gene to expression site association_stage qualifier,gene to expression site association_quantifier qualifier,gene to expression site association_subject,gene to expression site association_object,gene to expression site association_predicate\n",
      "biolink:GeneToDiseaseAssociation|id,iri,name,description,source,provided by,has attribute,predicate,object,relation,negated,qualifiers,publications,association_type,association_category,gene to disease association_subject,frequency qualifier,severity qualifier,onset qualifier\n",
      "biolink:GeneHasVariantThatContributesToDiseaseAssociation|id,iri,name,description,source,provided by,has attribute,predicate,object,relation,negated,qualifiers,publications,association_type,association_category,frequency qualifier,severity qualifier,onset qualifier,sequence variant qualifier,gene has variant that contributes to disease association_subject\n",
      "biolink:GeneAsAModelOfDiseaseAssociation|id,iri,name,description,source,provided by,has attribute,predicate,object,relation,negated,qualifiers,publications,association_type,association_category,frequency qualifier,severity qualifier,onset qualifier,gene as a model of disease association_subject\n",
      "biolink:MaterialSampleDerivationAssociation|id,iri,name,description,source,provided by,has attribute,relation,negated,qualifiers,publications,association_type,association_category,material sample derivation association_subject,material sample derivation association_object,material sample derivation association_predicate\n",
      "biolink:OrganismTaxonToEnvironmentAssociation|id,iri,name,description,source,provided by,has attribute,relation,negated,qualifiers,publications,association_type,association_category,organism taxon to environment association_subject,organism taxon to environment association_object,organism taxon to environment association_predicate\n",
      "biolink:GeneToPhenotypicFeatureAssociation|id,iri,name,description,source,provided by,has attribute,predicate,object,relation,negated,qualifiers,publications,association_type,association_category,gene to phenotypic feature association_subject,frequency qualifier,severity qualifier,onset qualifier,sex qualifier\n",
      "biolink:PopulationToPopulationAssociation|id,iri,name,description,source,provided by,has attribute,relation,negated,qualifiers,publications,association_type,association_category,population to population association_subject,population to population association_object,population to population association_predicate\n",
      "biolink:ContributorAssociation|id,iri,name,description,source,provided by,has attribute,relation,negated,publications,association_type,association_category,contributor association_subject,contributor association_predicate,contributor association_object,contributor association_qualifiers\n",
      "biolink:ChemicalToChemicalAssociation|id,iri,name,description,source,provided by,has attribute,subject,predicate,relation,negated,qualifiers,publications,association_type,association_category,chemical to chemical association_object\n",
      "biolink:ReactionToParticipantAssociation|id,iri,name,description,source,provided by,has attribute,predicate,relation,negated,qualifiers,publications,association_type,association_category,chemical to chemical association_object,stoichiometry,reaction direction,reaction side,reaction to participant association_subject\n",
      "biolink:ReactionToCatalystAssociation|id,iri,name,description,source,provided by,has attribute,predicate,relation,negated,qualifiers,publications,association_type,association_category,stoichiometry,reaction direction,reaction side,reaction to participant association_subject,reaction to catalyst association_object\n",
      "biolink:ChemicalToChemicalDerivationAssociation|id,iri,name,description,source,provided by,has attribute,relation,negated,qualifiers,publications,association_type,association_category,chemical to chemical derivation association_catalyst qualifier,chemical to chemical derivation association_subject,chemical to chemical derivation association_object,chemical to chemical derivation association_predicate\n",
      "biolink:AnatomicalEntityToAnatomicalEntityAssociation|id,iri,name,description,source,provided by,has attribute,predicate,relation,negated,qualifiers,publications,association_type,association_category,anatomical entity to anatomical entity association_subject,anatomical entity to anatomical entity association_object\n",
      "biolink:AnatomicalEntityToAnatomicalEntityPartOfAssociation|id,iri,name,description,source,provided by,has attribute,relation,negated,qualifiers,publications,association_type,association_category,anatomical entity to anatomical entity part of association_subject,anatomical entity to anatomical entity part of association_object,anatomical entity to anatomical entity part of association_predicate\n",
      "biolink:AnatomicalEntityToAnatomicalEntityOntogenicAssociation|id,iri,name,description,source,provided by,has attribute,relation,negated,qualifiers,publications,association_type,association_category,anatomical entity to anatomical entity ontogenic association_subject,anatomical entity to anatomical entity ontogenic association_object,anatomical entity to anatomical entity ontogenic association_predicate\n",
      "biolink:GeneToGeneAssociation|id,iri,name,description,source,provided by,has attribute,predicate,relation,negated,qualifiers,publications,association_type,association_category,gene to gene association_subject,gene to gene association_object\n",
      "biolink:PairwiseGeneToGeneInteraction|id,iri,name,description,source,provided by,has attribute,negated,qualifiers,publications,association_type,association_category,gene to gene association_subject,gene to gene association_object,pairwise gene to gene interaction_predicate,pairwise gene to gene interaction_relation\n",
      "biolink:PairwiseMolecularInteraction|iri,name,description,source,provided by,has attribute,negated,qualifiers,publications,association_type,association_category,interacting molecules category,pairwise molecular interaction_subject,pairwise molecular interaction_id,pairwise molecular interaction_predicate,pairwise molecular interaction_relation,pairwise molecular interaction_object\n",
      "biolink:GeneToGeneCoexpressionAssociation|id,iri,name,description,source,provided by,has attribute,relation,negated,qualifiers,publications,association_type,association_category,gene to gene association_subject,gene to gene association_object,gene to gene coexpression association_predicate,gene expression mixin_quantifier qualifier,expression site,stage qualifier,phenotypic state\n",
      "biolink:GeneToGeneHomologyAssociation|id,iri,name,description,source,provided by,has attribute,relation,negated,qualifiers,publications,association_type,association_category,gene to gene association_subject,gene to gene association_object,gene to gene homology association_predicate\n",
      "biolink:SequenceFeatureRelationship|id,iri,name,description,source,provided by,has attribute,predicate,relation,negated,qualifiers,publications,association_type,association_category,sequence feature relationship_subject,sequence feature relationship_object\n",
      "biolink:ExonToTranscriptRelationship|id,iri,name,description,source,provided by,has attribute,predicate,relation,negated,qualifiers,publications,association_type,association_category,exon to transcript relationship_subject,exon to transcript relationship_object\n",
      "biolink:GeneToGeneProductRelationship|id,iri,name,description,source,provided by,has attribute,relation,negated,qualifiers,publications,association_type,association_category,gene to gene product relationship_subject,gene to gene product relationship_object,gene to gene product relationship_predicate\n",
      "biolink:TranscriptToGeneRelationship|id,iri,name,description,source,provided by,has attribute,predicate,relation,negated,qualifiers,publications,association_type,association_category,transcript to gene relationship_subject,transcript to gene relationship_object\n",
      "biolink:DiseaseOrPhenotypicFeatureToLocationAssociation|id,iri,name,description,source,provided by,has attribute,subject,predicate,relation,negated,qualifiers,publications,association_type,association_category,disease or phenotypic feature to location association_object\n",
      "biolink:VariantToPopulationAssociation|id,iri,name,description,source,provided by,has attribute,predicate,relation,negated,qualifiers,publications,association_type,association_category,variant to population association_subject,variant to population association_object,variant to population association_has quotient,variant to population association_has count,variant to population association_has total,has percentage,frequency qualifier\n",
      "biolink:CellLineToDiseaseOrPhenotypicFeatureAssociation|id,iri,name,description,source,provided by,has attribute,predicate,object,relation,negated,qualifiers,publications,association_type,association_category,cell line to disease or phenotypic feature association_subject\n",
      "biolink:CellLineAsAModelOfDiseaseAssociation|id,iri,name,description,source,provided by,has attribute,predicate,object,relation,negated,qualifiers,publications,association_type,association_category,cell line as a model of disease association_subject,frequency qualifier,severity qualifier,onset qualifier\n",
      "biolink:ExposureEventToOutcomeAssociation|id,iri,name,description,source,provided by,has attribute,subject,predicate,object,relation,negated,qualifiers,publications,association_type,association_category,has population context,has temporal context\n",
      "biolink:OrganismalEntityAsAModelOfDiseaseAssociation|id,iri,name,description,source,provided by,has attribute,predicate,object,relation,negated,qualifiers,publications,association_type,association_category,organismal entity as a model of disease association_subject,frequency qualifier,severity qualifier,onset qualifier\n",
      "biolink:ChemicalToPathwayAssociation|id,iri,name,description,source,provided by,has attribute,subject,predicate,relation,negated,qualifiers,publications,association_type,association_category,chemical to pathway association_object\n",
      "biolink:ChemicalToGeneAssociation|id,iri,name,description,source,provided by,has attribute,subject,predicate,relation,negated,qualifiers,publications,association_type,association_category,chemical to gene association_object\n",
      "biolink:FunctionalAssociation|id,iri,name,description,source,provided by,has attribute,predicate,relation,negated,qualifiers,publications,association_type,association_category,functional association_subject,functional association_object\n",
      "biolink:GeneToGoTermAssociation|id,iri,name,description,source,provided by,has attribute,predicate,relation,negated,qualifiers,publications,association_type,association_category,gene to go term association_subject,gene to go term association_object\n",
      "biolink:MacromolecularMachineToBiologicalProcessAssociation|id,iri,name,description,source,provided by,has attribute,predicate,relation,negated,qualifiers,publications,association_type,association_category,functional association_subject,macromolecular machine to biological process association_object\n",
      "biolink:MacromolecularMachineToCellularComponentAssociation|id,iri,name,description,source,provided by,has attribute,predicate,relation,negated,qualifiers,publications,association_type,association_category,functional association_subject,macromolecular machine to cellular component association_object\n",
      "biolink:MacromolecularMachineToMolecularActivityAssociation|id,iri,name,description,source,provided by,has attribute,predicate,relation,negated,qualifiers,publications,association_type,association_category,functional association_subject,macromolecular machine to molecular activity association_object\n",
      "biolink:ExposureEventToPhenotypicFeatureAssociation|id,iri,name,description,source,provided by,has attribute,predicate,object,relation,negated,qualifiers,publications,association_type,association_category,exposure event to phenotypic feature association_subject,frequency qualifier,severity qualifier,onset qualifier,sex qualifier\n",
      "biolink:VariantToDiseaseAssociation|id,iri,name,description,source,provided by,has attribute,relation,negated,qualifiers,publications,association_type,association_category,variant to disease association_subject,variant to disease association_predicate,variant to disease association_object,frequency qualifier,severity qualifier,onset qualifier\n",
      "biolink:VariantAsAModelOfDiseaseAssociation|id,iri,name,description,source,provided by,has attribute,relation,negated,qualifiers,publications,association_type,association_category,variant to disease association_predicate,variant to disease association_object,frequency qualifier,severity qualifier,onset qualifier,variant as a model of disease association_subject\n",
      "biolink:MaterialSampleToDiseaseOrPhenotypicFeatureAssociation|id,iri,name,description,source,provided by,has attribute,subject,predicate,object,relation,negated,qualifiers,publications,association_type,association_category\n",
      "biolink:GenotypeToDiseaseAssociation|id,iri,name,description,source,provided by,has attribute,relation,negated,qualifiers,publications,association_type,association_category,genotype to disease association_subject,genotype to disease association_predicate,genotype to disease association_object,frequency qualifier,severity qualifier,onset qualifier\n",
      "biolink:GenotypeAsAModelOfDiseaseAssociation|id,iri,name,description,source,provided by,has attribute,relation,negated,qualifiers,publications,association_type,association_category,genotype to disease association_predicate,genotype to disease association_object,frequency qualifier,severity qualifier,onset qualifier,genotype as a model of disease association_subject\n",
      "biolink:GenotypeToPhenotypicFeatureAssociation|id,iri,name,description,source,provided by,has attribute,object,relation,negated,qualifiers,publications,association_type,association_category,genotype to phenotypic feature association_predicate,genotype to phenotypic feature association_subject,frequency qualifier,severity qualifier,onset qualifier,sex qualifier\n",
      "biolink:DiseaseToExposureEventAssociation|id,iri,name,description,source,provided by,has attribute,subject,predicate,object,relation,negated,qualifiers,publications,association_type,association_category\n",
      "biolink:VariantToGeneAssociation|id,iri,name,description,source,provided by,has attribute,subject,relation,negated,qualifiers,publications,association_type,association_category,variant to gene association_object,variant to gene association_predicate\n",
      "biolink:VariantToGeneExpressionAssociation|id,iri,name,description,source,provided by,has attribute,subject,relation,negated,qualifiers,publications,association_type,association_category,variant to gene association_object,variant to gene expression association_predicate,gene expression mixin_quantifier qualifier,expression site,stage qualifier,phenotypic state\n"
     ]
    }
   ],
   "source": [
    "associations = []\n",
    "\n",
    "for association in tk.get_all_associations(True):\n",
    "    slots = \"\"\n",
    "    for slot in tk.get_element(association).slots:\n",
    "        if slots == \"\":\n",
    "            slots = slot\n",
    "        else:\n",
    "            slots = slots+','+slot\n",
    "    print (association +'|'+slots)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c40743c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import requests\n",
    "from pprint import pprint\n",
    "import pandas as pd\n",
    "import bioregistry as br\n",
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "92a75b53",
   "metadata": {},
   "outputs": [],
   "source": [
    "reusabledata_url = 'https://raw.githubusercontent.com/sierra-moxon/reusabledata/master/data.json'\n",
    "sierra_url = 'https://raw.githubusercontent.com/sierra-moxon/reusabledata/add_infores/data-sources/compiled.json'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "4b8d3c24",
   "metadata": {},
   "outputs": [],
   "source": [
    "response = requests.get(reusabledata_url)\n",
    "sierra_response = requests.get(sierra_url)\n",
    "source_list = []\n",
    "source_dict = []\n",
    "infores_list = []\n",
    "mapped_licenses = []\n",
    "\n",
    "csv_infores = 'https://raw.githubusercontent.com/sierra-moxon/reusabledata/add_infores/infores.csv'\n",
    "\n",
    "for source in response.json():\n",
    "    source_list.append(source.get('source'))\n",
    "    bits = {\"id\":source.get('source'),\"license\":source.get('license'), \"license_type\":source.get('license-type'),\n",
    "            \"license_commentary\": source.get('license-commentary')}\n",
    "    source_dict.append(bits)\n",
    "    \n",
    "for ssource in sierra_response.json():\n",
    "    source_list.append(ssource.get('id'))\n",
    "    bits = {\"id\":ssource.get('id'),\"license\":ssource.get('license'), \"license_type\":ssource.get('license-type'),\n",
    "            \"license_commentary\": ssource.get('license-commentary')}\n",
    "    source_dict.append(bits)\n",
    "csv_infores_response = requests.get(csv_infores)\n",
    "with requests.Session() as s:\n",
    "    download = s.get(csv_infores)\n",
    "\n",
    "    decoded_content = download.content.decode('utf-8')\n",
    "\n",
    "    cr = csv.reader(decoded_content.splitlines(), delimiter=',')\n",
    "    my_list = list(cr)\n",
    "    for row in my_list:\n",
    "        infores_list.append(row[5])\n",
    "        infores_list.append(row[6].replace(\"infores:\",\"\"))\n",
    "\n",
    "intersection = set(source_list).intersection(infores_list)\n",
    "for item in intersection:\n",
    "    for mapper in source_dict:\n",
    "        if mapper.get(\"id\") == item:\n",
    "            mapped_licenses.append(mapper)\n",
    "\n",
    "keys = mapped_licenses[0].keys()\n",
    "with open('/tmp/mapped_licenses.csv', 'w', newline='')  as output_file:\n",
    "    dict_writer = csv.DictWriter(output_file, keys)\n",
    "    dict_writer.writeheader()\n",
    "    dict_writer.writerows(mapped_licenses)\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28429a2b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
